pkg/cli/userfile.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package cli
+
+import (
+	"context"
+	"database/sql/driver"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/errors"
+	"github.com/spf13/cobra"
+)
+
+const defaultQualifiedDBSchemaName = "defaultdb.public."
+
+var userFileUploadCmd = &cobra.Command{
+	Use:   "upload <source> <destination>",
+	Short: "Upload file from source to destination",
+	Long: `
+Uploads a file to the user scoped file storage using a SQL connection.
+`,
+	Args: cobra.MinimumNArgs(2),
+	RunE: maybeShoutError(runUserFileUpload),
+}
+
+func runUserFileUpload(cmd *cobra.Command, args []string) error {
+	conn, err := makeSQLClient("cockroach userfile", useDefaultDb)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	source := args[0]
+	destination := args[1]
+	reader, err := openUserFile(source)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	return uploadUserFile(ctx, conn, reader, destination)
+}
+
+func openUserFile(source string) (io.ReadCloser, error) {
+	f, err := os.Open(source)
+	if err != nil {
+		return nil, err
+	}
+	stat, err := f.Stat()
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get source file stats for %s", source)
+	}
+	if stat.IsDir() {
+		return nil, fmt.Errorf("source file %s is a directory, not a file", source)
+	}
+	return f, nil
+}
+
+func uploadUserFile(
+	ctx context.Context, conn *sqlConn, reader io.Reader, destination string,
+) error {
+	if err := conn.ensureConn(); err != nil {
+		return err
+	}
+
+	ex := conn.conn.(driver.ExecerContext)
+	if _, err := ex.ExecContext(ctx, `BEGIN`, nil); err != nil {
+		return err
+	}
+
+	// TODO(adityamaru): In the future we may want to allow users to specify a
+	// fully qualified db.schema.table where their underlying SQL file tables will
+	// be created. Enforcing the filepath to begin with a / allows for easy
+	// disambiguation between the qualified name and the filepath.
+	if !strings.HasPrefix(destination, "/") {
+		return errors.Newf("userfile upload destination path must begin with /")
+	}
+
+	// TODO(adityamaru): We reject any destination filepath's with `..` in them.
+	// This is because as the UserFileTableSystem is not a real file system, when
+	// you upload a file to a destination such as test/../../test.csv, we write
+	// its contents to a SQL table with filename set to test/../../test.csv. This
+	// is strange and we should come up with a better scheme of enforcing
+	// "sensible" filenames.
```

Comment:
I think Go URLs support path normalization per RFC3986. https://tools.ietf.org/html/rfc3986

Perhaps we could normalize a URL by parsing it, then reconstruct it and avoid some headaches?

https://golang.org/src/net/url/url.go

Reply:
I still think we’re making a lot of these “paths” that we don’t need to — These are not actual file system paths, they’re just string names. S3 or google cloud storage just treat the name as a string, with slashes simply a display-time concept used for grouping results in the GUI. `gsutil cp local-file gs://my_bucket/path/../to/./file` produces a file named “ path/../to/./file” and if it is good enough for google, it is good enough for me.

Reply:
>  if it is good enough for google, it is good enough for me.

- You say "S3 or GCS". What does S3 say?

- S3 provides a HTTP interface to access stored files in a web browser. What does the web browser say about these special paths? Do they get normalized?

- you're casually dismissing the "UX surprise" factor. Users will see this an ask questions. Wouldn't it be clearer to side-step these discussions preemptively?

Reply:
```[13:21:41] ~/go/src/github.com/cockroachdb/cockroach $ aws s3 cp README.md s3://roland-test-s3encryption/a/path/../bar
upload: ./README.md to s3://roland-test-s3encryption/a/path/../bar

[13:22:00] ~/go/src/github.com/cockroachdb/cockroach $ aws s3 ls s3://roland-test-s3encryption/a/
                           PRE path/

[13:22:10] ~/go/src/github.com/cockroachdb/cockroach $ aws s3 ls s3://roland-test-s3encryption/a/path/
                           PRE ../

[13:22:14] ~/go/src/github.com/cockroachdb/cockroach $ aws s3 ls s3://roland-test-s3encryption/a/path/../
2020-07-14 13:22:01       6494 bar```

Reply:
AWS HTTP API just treats the dots as part of the name. A GET request to "http://bucket.s3-website-us-east-1.amazonaws.com/a/path/../bar" will return that file, after a ` aws s3api put-object-acl --bucket bucket --key a/path/foo --acl public-read`, where, again, in the `key` we see dots are just dots.

Reply:
So, is it a bit surprising? yeah, it could be, to someone who expects it to behave like their shell. But given that userfile sits side-by-side with s3 and gcs in our storage options, I'm in favor of it behaving consistently with them. The interaction we're building here maps 1:1 with `aws s3 cp local remote` or gsutil cp local remote` and obviously they both came to the same conclusion: just treat the supplied remote name as a string key with no special rewriting rules, and let the user decide what to put in it.

Reply:
> The interaction we're building here maps 1:1 with

I hear you and as long as this is true, all the points you make (and the choices in the PR) are OK.

Except that I expect this not to be true all the time. Before we know it we'll want to mock the stores by using a filesystem-backed store instead. Or there's a new cloud provider with a different idea on the topic. 

More generally, I am not keen to look just at S3 and GCS and then rewrite 60 years of filesystem design history, while praying that the future will not re-align. This is ad-hoc revisionism and bound to bite us in the future.

Finally, this creates an inconsistency between the handling of s3/gcs-backed bulk operations, and that supported by `nodelocal://`. Because `nodelocal` does normalize its paths.

Reply:
If or when we choose to back this with a real file system, it'd be on us to ensure ensure we internally use file names that underlying filesystem handles, either via escaping of the user-provided name, hashing it, or a completely opaque identifier to which we maintain a mapping. I'm confident we could solve that if we needed to.

Yes, `nodelocal` normalizes paths, but if anything, I'd prefer it didn't -- we have to take extreme care when doing so to ensure that the post-normalization path abides by the base-path restriction. I'd be in favor of instead of checking if post-normalization we're still in the base, we instead simply required normalization to be a noop and thus rejected paths that used dot segments. Additionally, we control the semantics of `nodelocal` but we don't control the semantics of the cloud providers, so if we want to provide consistent semantics, then the fact `nodelocal` is the odd-one-out is what we should fix, not emulate.

Reply:
> If or when we choose to back this with a real file system, it'd be on us to ensure ensure we internally use file names that underlying filesystem handles, either via escaping of the user-provided name, hashing it, or a completely opaque identifier to which we maintain a mapping

Ok that works for me.

> So, is it a bit surprising? yeah, it could be, to someone who expects it to behave like their shell. But given that userfile sits side-by-side with s3 and gcs in our storage options, I'm in favor of it behaving consistently with them.

This does not address the fact that a user will expect `a/./b` and `a/b` to be equivalent, as well as `a/b/../c` and `a/c`.

I get that S3/GCS have a different opinion on the topic, but it's unlikely that the *database user* will know/understand this. I think it's on CockroachDB to "make data easy" and remove this source of surprise.

At the very least I'm in favor of blocking these patterns. I would be surprised if users actively _desire_ to support them, an I'd be willing to block this feature until the point we have evidence it's needed.

Reply:
>> If or when we choose to back this with a real file system, it'd be on us to ensure ensure we internally use file names that underlying filesystem handles, either via escaping of the user-provided name, hashing it, or a completely opaque identifier to which we maintain a mapping

>Ok that works for me.

I concur here. The keys for the blobs should have no relation to the underlying storage system.

> At the very least I'm in favor of blocking these patterns. I would be surprised if users actively desire to support them, an I'd be willing to block this feature until the point we have evidence it's needed.

Sounds good to me.